### ISSUE TITLE
Protect dummies from history --clear

### ISSUE THREAD

**Issue by c22** [fish-shell/fish-shell#3307]
I've been using fish for a while now and just got burned by running:

```
history --contains "some command" --clear
y
```

When I intended to run:
```
history --contains "some command" --delete
1
```

I've had a few close calls with this in the past and I'm sure new users might fall into the same trap.

Given the other command line options don't do anything when you run `--clear` perhaps we can add some additional warning or protection if someone attempts to run `--clear` with additional options.

**Comment by floam** [comment]
Makes sense to me.

**Comment by c22** [comment]
Great, I can possibly look at implementing some protection and send a PR. I am not sure what you mean by the `-t` option, I don't see that as a switch to `history`.

**Comment by floam** [comment]
`-t` isn't in 2.3.1 but it makes it show timestamps in the current version. 

(I checked my history and I appear to have used `history -t --clear` once. I had just done a `history -t` and was lazy and then added a `--clear` to that when I wanted to clear apparently.)

If you will try to patch this, be sure you are working with git master, a lot of work was done recently by @krader1961 to make the argument handling less insane. I'd venture a guess there's a chance he might even beat you to it if you're not very quick :smile:.

**Comment by c22** [comment]
Regarding the scope, I am not sure if anyone would ever legitimately want to clear their history and merge at the same time, but I see it less of a trap.

If we categorise `--search`/`--prefix`/`--contains` as "filters" then it makes sense to me to add a warning (or simply error out, as you said) if someone attempts to add a filter to `--clear`, that's really where the trap lies.

Further on the discussion of how things like `history --merge --clear` should work, does this syntax currently even do anything other than clear? Seems to me like `--clear` essentially short circuits anything else that the history command might achieve (though I haven't tested it). At that point it could be argued that specifying additional options to any of `history ( --merge | --save | --clear )` should throw an error.

**Comment by floam** [comment]
Huh, looking over the man page: I'm not sure it was wise to promote `--clear` with a short option (`-l`). Or `--save` which was supposed to be internal.

**Comment by floam** [comment]
>Further on the discussion of how things like history --merge --clear should work, does this syntax currently even do anything other than clear? 

On git `history --contains foo --clear` shows items matching foo, it doesn't clear. `history --clear --contains foo` is trying to clear.

**Comment by c22** [comment]
If you read the (2.3.1) man page the way I read it:

>    history ( --merge | --save | --clear )
>    history ( --search | --delete ) [ --prefix 'prefix string' | --contains 'search string' ]

It really starts to make sense that specifying any of `--merge`, `--save`, `--clear`, `--search` or `--delete` in the same invocation should simply error out, and specifying `--prefix` or `--contains` on the same invocation as a `--merge` `--save` or `--clear` should also produce an error. I am sure that's how other commands with similar syntax behave, but unfortunately I can't think of any examples right now.

As another example, current behaviour of `history --search "term" --delete` short circuits the `--delete` and `--delete` achieves nothing.

Display switches like the aforementioned `-t` could probably just be silently ignored.

**Comment by floam** [comment]
It does - this got better after 2.3.1:

```
history --merge --save
You cannot specify multiple commands: merge save
```

**Comment by c22** [comment]
Seems like things have definitely improved but could perhaps be further improved based on your comment above.
> history --clear --contains foo is trying to clear.
If most of the work has been done by @krader1961 I'll "watch this space" for now :)

**Comment by krader1961** [comment]
I can't reproduce the error you saw with `history --clear -t` but that is no doubt due to the improvements I recently merged to bring some sanity to the behavior of this command. As for your proposal, @c22, I agree that the operations which take no other arguments should error out if other arguments are provided; including `-t` or `--time` since that doesn't make any sense for those operations. I'm also inclined to require a more verbose response to the

```
Are you sure you want to clear history? (y/n)
```

prompt than a simple "y" given the catastrophic nature of the operation. In fact, there should probably be additional text before that prompt which makes it clear what will happen if the user answers in the affirmative.

Note that both the fish script and C++ code needs to be modified. If you're not comfortable doing that I'll take care of doing this otherwise I'll review your pull-request.

**Comment by floam** [comment]
Reopening instead of making a new issue: Related to this endeavor, I noticed certain short options were added to history since after the last release, for stuff we most likely really only want there to be long options for: `--save`, which is not intended to even be used by users, and `--clear`. Those should only work through long-options. 

Not every long option should necessarily have an accompanying short option, only those options used often enough that we want to enable a shortcut for them should be bestowed with a short option. Certainly dangerous commands or "don't use this" commands should require the long option to be used explicitly.
